crypto/provider_core.c

```diff
@@ -1451,13 +1450,28 @@ int ossl_provider_doall_activated(OSSL_LIB_CTX *ctx,
     for (curr++; curr < max; curr++) {
         OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(provs, curr);
 
-        provider_deactivate(prov, 0, 1);
+        if (!CRYPTO_DOWN_REF(&prov->activatecnt, &ref, prov->refcnt_lock)) {
+            ret = 0;
+            continue;
+        }
+        if (ref < 1) {
+            /*
+             * Looks like we need to deactivate properly. We could just have
+             * done this originally, but it involves taking a write lock so
+             * we avoid it. We up the count again and do a full deactivation
+             */
+            if (CRYPTO_UP_REF(&prov->activatecnt, &ref, prov->refcnt_lock) > 0)
+                provider_deactivate(prov, 0, 1);
+        }
```

Comment:
As commented above, assuming the documentation is correct, you could just have called provider_deactivate() and just take the write lock when you actually need it.

Reply:
provider_deactivate does a whole load of stuff we're not expecting to need in the normal straight forward case, so I'd prefer to avoid it.